pkg/resource/deploy/step_generator.go

```diff
@@ -690,20 +694,11 @@ func (sg *stepGenerator) diff(urn resource.URN, old, new *resource.State, oldInp
 		return plugin.DiffResult{Changes: plugin.DiffSome, ReplaceKeys: []resource.PropertyKey{"provider"}}, nil
 	}
 
-	// Workaround #1251: unexpected replaces.
-	//
-	// The legacy/desired behavior here is that if the provider-calculated inputs for a resource did not change,
-	// then the resource itself should not change. Unfortunately, we (correctly?) pass the entire current state
-	// of the resource to Diff, which includes calculated/output properties that may differ from those present
-	// in the input properties. This can cause unexpected diffs.
-	//
-	// For now, simply apply the legacy diffing behavior before deferring to the provider.
-	if oldInputs.DeepEquals(newInputs) {
-		return plugin.DiffResult{Changes: plugin.DiffNone}, nil
-	}
-
 	// If there is no provider for this resource, simply return a "diffs exist" result.
 	if prov == nil {
```

Comment:
For my own edification - "no provider for this resource" means it was a component resource, or are there custom resources that don't have providers?

Reply:
This should only happen in the case of component resources.